packages/accounts/src/smart_account/storage.rs

```diff
@@ -0,0 +1,999 @@
+    CreateContract(BytesN<32>),
+}
+
+/// Metadata for a context rule.
+#[contracttype]
+#[derive(Clone, Debug, PartialEq)]
+pub struct Meta {
+    /// Human-readable name for the context rule.
+    pub name: String,
+    /// The type of context this rule applies to.
+    pub context_type: ContextRuleType,
+    /// Optional expiration ledger sequence for the rule.
+    pub valid_until: Option<u32>,
+}
+
+/// A complete context rule defining authorization requirements.
+#[contracttype]
+#[derive(Clone, Debug, PartialEq)]
+pub struct ContextRule {
+    /// Unique identifier for the context rule.
+    pub id: u32,
+    /// The type of context this rule applies to.
+    pub context_type: ContextRuleType,
+    /// Human-readable name for the context rule.
+    pub name: String,
+    /// List of signers authorized by this rule.
+    pub signers: Vec<Signer>,
+    /// List of policy contracts that must be satisfied.
+    pub policies: Vec<Address>,
+    /// Optional expiration ledger sequence for the rule.
+    pub valid_until: Option<u32>,
+}
+
+// ################## QUERY STATE ##################
+
+/// Retrieves a complete context rule by its ID.
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+/// * `id` - The unique identifier of the context rule.
+///
+/// # Errors
+///
+/// * [`SmartAccountError::ContextRuleNotFound`] - When the context rule with
+///   the specified ID does not exist.
+pub fn get_context_rule(e: &Env, id: u32) -> ContextRule {
+    let meta_key = SmartAccountStorageKey::Meta(id);
+    let meta: Meta = get_persistent_entry(e, &meta_key)
+        .unwrap_or_else(|| panic_with_error!(e, SmartAccountError::ContextRuleNotFound));
+
+    let signers_key = SmartAccountStorageKey::Signers(id);
+    let signers: Vec<Signer> = get_persistent_entry(e, &signers_key).unwrap_or(Vec::new(e));
+
+    let policies_key = SmartAccountStorageKey::Policies(id);
+    let policies: Vec<Address> = get_persistent_entry(e, &policies_key).unwrap_or(Vec::new(e));
+
+    ContextRule {
+        id,
+        context_type: meta.context_type,
+        name: meta.name,
+        signers,
+        policies,
+        valid_until: meta.valid_until,
+    }
+}
+
+/// Retrieves all context rules of a specific context type. Returns a vector of
+/// all context rules matching the specified type, including expired rules.
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+/// * `context_rule_type` - The type of context rules to retrieve.
+pub fn get_context_rules(e: &Env, context_rule_type: &ContextRuleType) -> Vec<ContextRule> {
+    let ids_key = SmartAccountStorageKey::Ids(context_rule_type.clone());
+    let ids: Vec<u32> = get_persistent_entry(e, &ids_key).unwrap_or(Vec::new(e));
+
+    let mut rules = Vec::new(e);
+    for id in ids.iter() {
+        if e.storage().persistent().has(&SmartAccountStorageKey::Meta(id)) {
+            rules.push_back(get_context_rule(e, id));
+        }
```

Comment:
It is quite inefficient to query the database twice with the same key:
- one for `has()`
- one for `get_context_rule()`

One option may be to write a low-level helper function, that is same with `get_context_rule`, but returns `Option` instead. 

This way:
- `get_context_rule` can utilize this, and if the return value is `None`, it can still panic.
- `get_context_rules` can also utilize this, and simply can use `unwrap_or()` like logic instead of querying the storage 2 times